In [1]:
import pandas as pd

In [2]:
#lecture du jeu de données , low_memory pour éviter les erreurs dans les colonnes avec mixed data types
df = pd.read_csv('..\\..\\accepted.csv', low_memory=False)

In [3]:
# on jette un coup d'oeil sur la forme du dataset 
print(df.shape)

(2260701, 151)


On constate qu'il y a 151 colonnes, on va tenter de prendre un sous ensemble de colonnes seulement, les colonnes clé.

In [4]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  68407277        NaN     3600.0       3600.0           3600.0   36 months   
1  68355089        NaN    24700.0      24700.0          24700.0   36 months   
2  68341763        NaN    20000.0      20000.0          20000.0   60 months   
3  66310712        NaN    35000.0      35000.0          35000.0   60 months   
4  68476807        NaN    10400.0      10400.0          10400.0   60 months   

   int_rate  installment grade sub_grade  ... hardship_payoff_balance_amount  \
0     13.99       123.03     C        C4  ...                            NaN   
1     11.99       820.28     C        C1  ...                            NaN   
2     10.78       432.66     B        B4  ...                            NaN   
3     14.85       829.90     C        C5  ...                            NaN   
4     22.45       289.91     F        F1  ...                            NaN   

  hardship_last_payment_amount disbursement_method  debt_settlement_flag  \
0                          NaN                Cash                     N   
1                          NaN                Cash                     N   
2                          NaN                Cash                     N   
3                          NaN                Cash                     N   
4                          NaN                Cash                     N   

  debt_settlement_flag_date settlement_status settlement_date  \
0                       NaN               NaN             NaN   
1                       NaN               NaN             NaN   
2                       NaN               NaN             NaN   
3                       NaN               NaN             NaN   
4                       NaN               NaN             NaN   

  settlement_amount settlement_percentage settlement_term  
0               NaN                   NaN             NaN  
1               NaN                   NaN             NaN  
2               NaN                   NaN             NaN  
3               NaN                   NaN             NaN  
4               NaN                   NaN             NaN  

[5 rows x 151 columns]

In [5]:
threshold = int(0.3*len(df.columns))
df.dropna(thresh=threshold)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0        68407277        NaN     3600.0       3600.0           3600.0   
1        68355089        NaN    24700.0      24700.0          24700.0   
2        68341763        NaN    20000.0      20000.0          20000.0   
3        66310712        NaN    35000.0      35000.0          35000.0   
4        68476807        NaN    10400.0      10400.0          10400.0   
...           ...        ...        ...          ...              ...   
2260694  89885898        NaN    24000.0      24000.0          24000.0   
2260695  88977788        NaN    24000.0      24000.0          24000.0   
2260696  88985880        NaN    40000.0      40000.0          40000.0   
2260697  88224441        NaN    24000.0      24000.0          24000.0   
2260698  88215728        NaN    14000.0      14000.0          14000.0   

               term  int_rate  installment grade sub_grade  ...  \
0         36 months     13.99       123.03     C        C4  ...   
1         36 months     11.99       820.28     C        C1  ...   
2         60 months     10.78       432.66     B        B4  ...   
3         60 months     14.85       829.90     C        C5  ...   
4         60 months     22.45       289.91     F        F1  ...   
...             ...       ...          ...   ...       ...  ...   
2260694   60 months     12.79       543.50     C        C1  ...   
2260695   60 months     10.49       515.74     B        B3  ...   
2260696   60 months     10.49       859.56     B        B3  ...   
2260697   60 months     14.49       564.56     C        C4  ...   
2260698   60 months     14.49       329.33     C        C4  ...   

        hardship_payoff_balance_amount hardship_last_payment_amount  \
0                                  NaN                          NaN   
1                                  NaN                          NaN   
2                                  NaN                          NaN   
3                                  NaN                          NaN   
4                                  NaN                          NaN   
...                                ...                          ...   
2260694                            NaN                          NaN   
2260695                            NaN                          NaN   
2260696                            NaN                          NaN   
2260697                            NaN                          NaN   
2260698                            NaN                          NaN   

        disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
0                      Cash                     N                       NaN   
1                      Cash                     N                       NaN   
2                      Cash                     N                       NaN   
3                      Cash                     N                       NaN   
4                      Cash                     N                       NaN   
...                     ...                   ...                       ...   
2260694                Cash                     N                       NaN   
2260695                Cash                     N                       NaN   
2260696                Cash                     N                       NaN   
2260697                Cash                     Y                  Mar-2019   
2260698                Cash                     N                       NaN   

        settlement_status settlement_date settlement_amount  \
0                     NaN             NaN               NaN   
1                     NaN             NaN               NaN   
2                     NaN             NaN               NaN   
3                     NaN             NaN               NaN   
4                     NaN             NaN               NaN   
...                   ...             ...               ...   
2260694               NaN             NaN               NaN   
2260695               NaN  

Data Cleaning

In [6]:
#on supprime les lignes qui contiennent plus de 70% de valeurs null
df = df.dropna(thresh=threshold)
df.shape

(2260668, 151)

In [7]:
df[df['member_id'].notna()]

Empty DataFrame
Columns: [id, member_id, loan_amnt, funded_amnt, funded_amnt_inv, term, int_rate, installment, grade, sub_grade, emp_title, emp_length, home_ownership, annual_inc, verification_status, issue_d, loan_status, pymnt_plan, url, desc, purpose, title, zip_code, addr_state, dti, delinq_2yrs, earliest_cr_line, fico_range_low, fico_range_high, inq_last_6mths, mths_since_last_delinq, mths_since_last_record, open_acc, pub_rec, revol_bal, revol_util, total_acc, initial_list_status, out_prncp, out_prncp_inv, total_pymnt, total_pymnt_inv, total_rec_prncp, total_rec_int, total_rec_late_fee, recoveries, collection_recovery_fee, last_pymnt_d, last_pymnt_amnt, next_pymnt_d, last_credit_pull_d, last_fico_range_high, last_fico_range_low, collections_12_mths_ex_med, mths_since_last_major_derog, policy_code, application_type, annual_inc_joint, dti_joint, verification_status_joint, acc_now_delinq, tot_coll_amt, tot_cur_bal, open_acc_6m, open_act_il, open_il_12m, open_il_24m, mths_since_rcnt_il, total_bal_il, il_util, open_rv_12m, open_rv_24m, max_bal_bc, all_util, total_rev_hi_lim, inq_fi, total_cu_tl, inq_last_12m, acc_open_past_24mths, avg_cur_bal, bc_open_to_buy, bc_util, chargeoff_within_12_mths, delinq_amnt, mo_sin_old_il_acct, mo_sin_old_rev_tl_op, mo_sin_rcnt_rev_tl_op, mo_sin_rcnt_tl, mort_acc, mths_since_recent_bc, mths_since_recent_bc_dlq, mths_since_recent_inq, mths_since_recent_revol_delinq, num_accts_ever_120_pd, num_actv_bc_tl, num_actv_rev_tl, num_bc_sats, num_bc_tl, num_il_tl, num_op_rev_tl, ...]
Index: []

[0 rows x 151 columns]

In [8]:
#on se débarasse de member_id puisqu'elle ne contient que des valeurs NULL
df=df.drop(columns=['member_id'])

In [9]:
#on jette toutes les colonnes qui ont 100% des valeur NaN
df_reduced = df.dropna(axis=1, how='all')
df_reduced.shape


(2260668, 150)

In [10]:
df_reduced.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2260668 entries, 0 to 2260698
Columns: 150 entries, id to settlement_term
dtypes: float64(112), object(38)
memory usage: 2.5+ GB


In [11]:
percentage_missing = df_reduced.isnull().sum() / len(df_reduced) * 100

In [12]:
print(percentage_missing)

id                        0.000000
loan_amnt                 0.000000
funded_amnt               0.000000
funded_amnt_inv           0.000000
term                      0.000000
                           ...    
settlement_status        98.485138
settlement_date          98.485138
settlement_amount        98.485138
settlement_percentage    98.485138
settlement_term          98.485138
Length: 150, dtype: float64


In [13]:
df_neat = pd.DataFrame(columns=df_reduced.columns, index=None)
df_neat.loc['percentage']=percentage_missing.values

In [14]:
pd.set_option('display.max_columns', None)



In [15]:
#On ne garde queles colonnes qui contiennent moins de 20% de NaN
colonnes_gardes = df_neat.columns[percentage_missing < 30].to_list()


In [16]:
print(colonnes_gardes)

['id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low', 'fico_range_high', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'total_rev_hi_lim', 'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct

In [17]:
print("nombre de colonnes avant :" , len(df_neat.columns))
print("nombre de colonnes après :" , len(colonnes_gardes))

nombre de colonnes avant : 150
nombre de colonnes après : 93


ce n'est toujours pas assez, on a besoin de plus de nettoyage

In [18]:
df_reduced=df_reduced[colonnes_gardes]

In [19]:
df_reduced['loan_status'].unique()

array(['Fully Paid', 'Current', 'Charged Off', 'In Grace Period',
       'Late (31-120 days)', 'Late (16-30 days)', 'Default',
       'Does not meet the credit policy. Status:Fully Paid',
       'Does not meet the credit policy. Status:Charged Off'],
      dtype=object)



Les variables suivantes sont utilisées pour prédire l'accord d'un prêt bancaire :

### Caractéristiques du prêt
- **loan_amnt** : Montant demandé 
- **term** : Durée du prêt (ex. "36 months").
- **int_rate** : Taux d’intérêt (%).
- **installment** : Mensualité (en $).
- **grade** : Note de risque (A à G).
- **sub_grade** : Sous-note détaillée (ex. A1).

### Situation personnelle
- **emp_length** : Ancienneté dans l’emploi (ex. "10+ years").
- **home_ownership** : Statut logement (ex. "RENT").

### Revenu et endettement
- **annual_inc** : Revenu annuel (en $).
- **verification_status** : Vérification revenu (ex. "Verified").
- **dti** : Ratio dette/revenu (%).

### Historique de crédit
- **delinq_2yrs** : Défauts sur 2 ans.
- **inq_last_6mths** : Demandes récentes de crédit.
- **open_acc** : Comptes actifs.
- **pub_rec** : Incidents publics.
- **revol_bal** : Solde crédits renouvelables (en $).
- **revol_util** : Utilisation crédits renouvelables (%).
- **total_acc** : Total comptes.

### Prêt en cours
- **last_pymnt_amnt** : Dernier paiement (en $).

### Cible
- **loan_status** : Statut prêt (ex. "Fully Paid"). À binariser pour la prédiction.

In [20]:
nos_features=['loan_amnt','term', 'int_rate', 'installment', 'grade', 'sub_grade','emp_length','home_ownership',
                'annual_inc','verification_status','purpose','dti','delinq_2yrs','inq_last_6mths','open_acc',
                'pub_rec','revol_bal','revol_util','total_acc','last_pymnt_amnt','loan_status']


In [21]:
len(nos_features)

21

In [22]:
df_reduced=df_reduced[nos_features]

In [23]:
df_reduced.describe().T

count          mean            std     min       25%  \
loan_amnt        2260668.0  15046.931228    9190.245488  500.00   8000.00   
int_rate         2260668.0     13.092829       4.832138    5.31      9.49   
installment      2260668.0    445.806823     267.173535    4.93    251.65   
annual_inc       2260664.0  77992.428687  112696.199574    0.00  46000.00   
dti              2258957.0     18.824196      14.183329   -1.00     11.89   
delinq_2yrs      2260639.0      0.306879       0.867230    0.00      0.00   
inq_last_6mths   2260638.0      0.576835       0.885963    0.00      0.00   
open_acc         2260639.0     11.612402       5.640861    0.00      8.00   
pub_rec          2260639.0      0.197528       0.570515    0.00      0.00   
revol_bal        2260668.0  16658.458078   22948.305028    0.00   5950.00   
revol_util       2258866.0     50.337696      24.713073    0.00     31.50   
total_acc        2260639.0     24.162552      11.987528    1.00     15.00   
last_pymnt_amnt  2260668.0   3429.345942    6018.247582    0.00    310.33   

                      50%       75%           max  
loan_amnt        12900.00  20000.00  4.000000e+04  
int_rate            12.62     15.99  3.099000e+01  
installment        377.99    593.32  1.719830e+03  
annual_inc       65000.00  93000.00  1.100000e+08  
dti                 17.84     24.49  9.990000e+02  
delinq_2yrs          0.00      0.00  5.800000e+01  
inq_last_6mths       0.00      1.00  3.300000e+01  
open_acc            11.00     14.00  1.010000e+02  
pub_rec              0.00      0.00  8.600000e+01  
revol_bal        11324.00  20246.00  2.904836e+06  
revol_util          50.30     69.40  8.923000e+02  
total_acc           22.00     31.00  1.760000e+02  
last_pymnt_amnt    600.57   3743.75  4.219205e+04

In [59]:
df_reduced.head()

loan_amnt  term  int_rate  installment home_ownership  annual_inc  \
0     3600.0  36.0     13.99       123.03       MORTGAGE     55000.0   
1    24700.0  36.0     11.99       820.28       MORTGAGE     65000.0   
2    20000.0  60.0     10.78       432.66       MORTGAGE     63000.0   
3    35000.0  60.0     14.85       829.90       MORTGAGE    110000.0   
4    10400.0  60.0     22.45       289.91       MORTGAGE    104433.0   

  verification_status             purpose    dti  delinq_2yrs  inq_last_6mths  \
0        Not Verified  debt_consolidation   5.91          0.0             1.0   
1        Not Verified      small_business  16.06          1.0             4.0   
2        Not Verified    home_improvement  10.78          0.0             0.0   
3     Source Verified  debt_consolidation  17.06          0.0             0.0   
4     Source Verified      major_purchase  25.37          1.0             3.0   

   open_acc  pub_rec  revol_bal  revol_util  total_acc  last_pymnt_amnt  \
0       7.0      0.0     2765.0        29.7       13.0           122.67   
1      22.0      0.0    21470.0        19.2       38.0           926.35   
2       6.0      0.0     7869.0        56.2       18.0         15813.30   
3      13.0      0.0     7802.0        11.6       17.0           829.90   
4      12.0      0.0    21929.0        64.5       35.0         10128.96   

   loan_status  
0          1.0  
1          1.0  
2          1.0  
3          0.0  
4          1.0

model

In [41]:
df_reduced["term"] = df_reduced["term"].str.extract(r'(\d+)').astype(float)


In [47]:
df_reduced.drop(columns="grade", inplace=True)

In [49]:
df_reduced.drop(columns="sub_grade", inplace=True)

In [50]:
df_reduced.drop(columns="emp_length", inplace=True)

In [89]:
df_reduced.drop(columns="installment", inplace=True)

In [56]:
df_reduced.loan_status.value_counts()

loan_status
Fully Paid                                             1076751
Current                                                 878317
Charged Off                                             268559
Late (31-120 days)                                       21467
In Grace Period                                           8436
Late (16-30 days)                                         4349
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     40
Name: count, dtype: int64

In [ ]:
df_reduced["loan_status"] = (df_reduced["loan_status"] == "Fully Paid").replace({True : 1.0, False : 0.0})

In [58]:
df_reduced.loan_status.value_counts()

loan_status
0.0    1183917
1.0    1076751
Name: count, dtype: int64

In [ ]:
dummies = pd.get_dummies(df_reduced["home_ownership"])[["RENT", "OWN"]].replace({True : 1.0, False : 0.0})
df_reduced = df_reduced.drop(columns=["home_ownership"])

df_reduced = pd.concat([df_reduced, dummies], axis=1)

C:\Users\yacin\AppData\Local\Temp\ipykernel_13184\2664449456.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dummies = pd.get_dummies(df_reduced["home_ownership"])[["RENT", "OWN"]].replace({True : 1.0, False : 0.0})


In [ ]:
dummies = pd.get_dummies(df_reduced["verification_status"])[["Source Verified", "Verified"]].replace({True : 1.0, False : 0.0})
df_reduced = df_reduced.drop(columns=["verification_status"])

df_reduced = pd.concat([df_reduced, dummies], axis=1)

C:\Users\yacin\AppData\Local\Temp\ipykernel_13184\2955468166.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dummies = pd.get_dummies(df_reduced["verification_status"])[["Source Verified", "Verified"]].replace({True : 1.0, False : 0.0})


In [82]:
df_reduced.purpose.value_counts()

purpose
debt_consolidation    1277877
credit_card            516971
home_improvement       150457
other                  139440
major_purchase          50445
medical                 27488
small_business          24689
car                     24013
vacation                15525
moving                  15403
house                   14136
wedding                  2355
renewable_energy         1445
educational               424
Name: count, dtype: int64

In [ ]:
selected_purposes = [
    "credit_card",
    "debt_consolidation",
    "home_improvement",
    "major_purchase",
    "medical",
    "other",
    "small_business"
]

dummies = pd.get_dummies(df_reduced["purpose"])[selected_purposes].replace({True: 1.0, False: 0.0})

df_reduced = df_reduced.drop(columns=["purpose"])

df_reduced = pd.concat([df_reduced, dummies], axis=1)


C:\Users\yacin\AppData\Local\Temp\ipykernel_13184\795537552.py:12: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dummies = pd.get_dummies(df_reduced["purpose"])[selected_purposes].replace({True: 1.0, False: 0.0})


entrainement

In [133]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler


X = df_reduced.drop(columns=["loan_status"]) 
y = df_reduced["loan_status"]  

X = X.dropna()
y = y[X.index] 

scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

logistic_model = LogisticRegression(max_iter=1000)
logistic_model.fit(X_train, y_train)

y_pred = logistic_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.4f}")


Accuracy: 0.8527


In [134]:
from sklearn.metrics import confusion_matrix, classification_report

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[230880   5729]
 [ 60752 154071]]

Classification Report:
              precision    recall  f1-score   support

         0.0       0.79      0.98      0.87    236609
         1.0       0.96      0.72      0.82    214823

    accuracy                           0.85    451432
   macro avg       0.88      0.85      0.85    451432
weighted avg       0.87      0.85      0.85    451432



on pickle le model

In [135]:
import os
import pickle

# Spécifier le répertoire où le modèle sera sauvegardé
model_dir = '..\\app\\models'

# Sauvegarder le modèle dans un fichier avec pickle
model_filename = os.path.join(model_dir, 'logistic_model.pkl')
with open(model_filename, 'wb') as file:
    pickle.dump(logistic_model, file)

print(f"Modèle sauvegardé dans : {model_filename}")

Modèle sauvegardé dans : ..\app\models\logistic_model.pkl


In [136]:
df_reduced[df_reduced.loan_status == 1.0].head()

loan_amnt  term  int_rate  annual_inc    dti  delinq_2yrs  inq_last_6mths  \
0     3600.0  36.0     13.99     55000.0   5.91          0.0             1.0   
1    24700.0  36.0     11.99     65000.0  16.06          1.0             4.0   
2    20000.0  60.0     10.78     63000.0  10.78          0.0             0.0   
4    10400.0  60.0     22.45    104433.0  25.37          1.0             3.0   
5    11950.0  36.0     13.44     34000.0  10.20          0.0             0.0   

   open_acc  pub_rec  revol_bal  revol_util  total_acc  last_pymnt_amnt  \
0       7.0      0.0     2765.0        29.7       13.0           122.67   
1      22.0      0.0    21470.0        19.2       38.0           926.35   
2       6.0      0.0     7869.0        56.2       18.0         15813.30   
4      12.0      0.0    21929.0        64.5       35.0         10128.96   
5       5.0      0.0     8822.0        68.4        6.0          7653.56   

   loan_status  RENT  OWN  Source Verified  Verified  credit_card  \
0          1.0   0.0  0.0              0.0       0.0          0.0   
1          1.0   0.0  0.0              0.0       0.0          0.0   
2          1.0   0.0  0.0              0.0       0.0          0.0   
4          1.0   0.0  0.0              1.0       0.0          0.0   
5          1.0   1.0  0.0              1.0       0.0          0.0   

   debt_consolidation  home_improvement  major_purchase  medical  other  \
0                 1.0               0.0             0.0      0.0    0.0   
1                 0.0               0.0             0.0      0.0    0.0   
2                 0.0               1.0             0.0      0.0    0.0   
4                 0.0               0.0             1.0      0.0    0.0   
5                 1.0               0.0             0.0      0.0    0.0   

   small_business  
0             0.0  
1             1.0  
2             0.0  
4             0.0  
5             0.0

In [139]:
df_reduced[df_reduced.loan_status == 0.0].head()

loan_amnt  term  int_rate  annual_inc    dti  delinq_2yrs  inq_last_6mths  \
3     35000.0  60.0     14.85    110000.0  17.06          0.0             0.0   
10    22400.0  60.0     12.88     95000.0  22.98          0.0             0.0   
11    16000.0  60.0     12.88     70000.0  26.40          0.0             0.0   
13    18000.0  60.0     19.48    150000.0   9.39          0.0             1.0   
18    23000.0  60.0      8.49     64000.0  18.28          0.0             0.0   

    open_acc  pub_rec  revol_bal  revol_util  total_acc  last_pymnt_amnt  \
3       13.0      0.0     7802.0        11.6       17.0           829.90   
10      16.0      0.0    13028.0        32.7       35.0           508.30   
11      13.0      0.0    28705.0        56.3       29.0           363.07   
13      18.0      1.0    14052.0        40.7       27.0           471.70   
18      21.0      0.0    24940.0        52.7       33.0           471.77   

    loan_status  RENT  OWN  Source Verified  Verified  credit_card  \
3           0.0   0.0  0.0              1.0       0.0          0.0   
10          0.0   0.0  0.0              0.0       0.0          0.0   
11          0.0   0.0  0.0              0.0       0.0          0.0   
13          0.0   1.0  0.0              0.0       0.0          0.0   
18          0.0   1.0  0.0              0.0       0.0          1.0   

    debt_consolidation  home_improvement  major_purchase  medical  other  \
3                  1.0               0.0             0.0      0.0    0.0   
10                 1.0               0.0             0.0      0.0    0.0   
11                 1.0               0.0             0.0      0.0    0.0   
13                 1.0               0.0             0.0      0.0    0.0   
18                 0.0               0.0             0.0      0.0    0.0   

    small_business  
3              0.0  
10             0.0  
11             0.0  
13             0.0  
18             0.0

sa nous aidera pour tester les deux cas

In [141]:
import numpy as np
input_data_5 = [11950.0, 36.0, 13.44, 34000.0, 10.20, 0.0, 0.0, 5.0, 0.0, 8822.0, 68.4, 6.0, 7653.56, 1.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]

input_data_reshaped = np.array(input_data_5).reshape(1, -1)

prediction = logistic_model.predict(input_data_reshaped)[0]

print(f"Prédiction pour l'entrée 5: {prediction}")


Prédiction pour l'entrée 5: 1.0


In [142]:
import numpy as np

input_data_3 = [35000.0, 60.0, 14.85, 110000.0, 17.06, 0.0, 0.0, 13.0, 0.0, 7802.0, 11.6, 17.0, 829.90, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0]


input_data_reshaped = np.array(input_data_3).reshape(1, -1)

prediction = logistic_model.predict(input_data_reshaped)[0]

print(f"Prédiction pour l'entrée 3: {prediction}")

Prédiction pour l'entrée 3: 0.0


pour conseiller on prend la moyenne des client accordé

In [143]:
filtered_df = df_reduced[df_reduced['loan_status'] == 1]
mean_values = filtered_df.mean()


In [144]:
mean_df = pd.DataFrame(mean_values).transpose()

In [145]:
with open('..\\app\\models\\meanpos.pkl', 'wb') as f:
    pickle.dump(mean_values.to_dict(), f)

In [ ]:
model_scaler = os.path.join("..\\app\\models\\scaler.pkl")
with open(model_scaler, 'wb') as f:
    pickle.dump(scaler, f)

In [ ]:
import shap
import cloudpickle
explainer = shap.Explainer(logistic_model, X_train)
with open('..\\app\\models\\explainer.pkl', 'wb') as f:
    cloudpickle.dump(explainer, f)

c:\Users\yacin\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
